pkg/storage/stores/shipper/gateway_client.go

```diff
@@ -136,6 +170,56 @@ func (s *GatewayClient) doQueries(ctx context.Context, queries []chunk.IndexQuer
 	return nil
 }
 
+// ringModeDoQueries prepares an index query to be sent to the Index Gateway, and then sends it
+// using the clientDoQueries implementation.
+//
+// The preparation and sending phase includes:
+// 1. Extracting the tenant name from the query.
+// 2. Fetching different Index Gateway instances assigned to the extracted tenant.
+// 3. Iterating in parallel over all fetched Index Gateway instances, getting their gRPC connections
+//  from the pool and invoking clientDoQueries using their client.
+func (s *GatewayClient) ringModeDoQueries(ctx context.Context, gatewayQueries []*indexgatewaypb.IndexQuery, queryKeyQueryMap map[string]chunk.IndexQuery, callback chunk.QueryPagesCallback) error {
+	// TODO: should we keep iterating if an error occur?
+	for _, q := range gatewayQueries {
+		bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
+
+		splittedHashed := strings.Split(q.HashValue, ":")
+		if len(splittedHashed) < 2 {
+			level.Error(util_log.Logger).Log("msg", "query hash value in incorrect format", "hash_value", q.HashValue)
+			continue
+		}
+
+		// tenant is the second parameter.
+		tenantID := splittedHashed[1]
+		key := util.TokenFor(tenantID, "" /* labels */)
+		rs, err := s.ring.Get(key, ring.WriteNoExtend, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			level.Error(util_log.Logger).Log("msg", "failed to get index key", "error", err)
+			continue
+		}
+
+		_, err = rs.Do(ctx, 0 /* delay, TODO: what is it for? */, func(ctx context.Context, instanceDesc *ring.InstanceDesc) (interface{}, error) {
+			genericClient, err := s.pool.GetClientFor(instanceDesc.Addr)
+			if err != nil {
+				return nil, errors.Wrap(err, fmt.Sprintf("get client for instance %s", instanceDesc.Addr))
+			}
+
+			client := (genericClient.(indexgatewaypb.IndexGatewayClient))
+			// TODO: does this works as a hedging request?
+			if err := s.clientDoQueries(ctx, gatewayQueries, queryKeyQueryMap, callback, client); err != nil {
```

Comment:
We are doing all the `n` queries `n*n` times.

Reply:
Great catch, thanks. I refactored it to not iterate over all queries anymore (by extracting the tenant from the ctx).

Reply:
Done. PTAL?